.github/workflows/build.yml

```
@@ -9,34 +9,64 @@ jobs:
   build:
     name: Build
     runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        java_version: ['11', '16']
     steps:
+      # Check out the project
       - uses: actions/checkout@v2
         with:
           fetch-depth: 0  # Shallow clones should be disabled for a better relevancy of analysis
-      - name: Set up JDK 11
+
+      # Setup the version of Java
+      - name: Set up JDK ${{ matrix.java_version }}
         uses: actions/setup-java@v1
         with:
-          java-version: 11
+          java-version: ${{ matrix.java_version }}
+
+      # Cache all the things
       - name: Cache SonarCloud packages
-        uses: actions/cache@v1
+        uses: actions/cache@v2.1.4
+        if: ${{ env.SONAR_TOKEN != null && env.SONAR_TOKEN != '' && matrix.java_version == '11' }}
+        env:
+          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
         with:
           path: ~/.sonar/cache
           key: ${{ runner.os }}-sonar
           restore-keys: ${{ runner.os }}-sonar
+
       - name: Cache Maven packages
-        uses: actions/cache@v1
+        uses: actions/cache@v2.1.4
         with:
           path: ~/.m2
           key: ${{ runner.os }}-m2-${{ hashFiles('**/pom.xml') }}
           restore-keys: ${{ runner.os }}-m2
+
       - name: Cache Embedded Mongo files
-        uses: actions/cache@v1
+        uses: actions/cache@v2.1.4
         with:
           path: ~/.embedmongo
           key: ${{ runner.os }}-embedmongo
           restore-keys: ${{ runner.os }}-embedmongo
-      - name: Build and analyze
+
+      # Compile the project
+      - name: Build
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}  # Needed to get PR information, if any
+        run: mvn -B -V compile
+
+      # Run tests
+      - name: Run tests
+        if: ${{ env.SONAR_TOKEN == null || env.SONAR_TOKEN == '' || matrix.java_version != '11' }}
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}  # Needed to get PR information, if any
+        run: mvn -B -V verify
+
+      # Run Sonar Analysis
```

similar to above, suggest something like:

```suggestion
      # Run Sonar Analysis (on Java version 11 only)
```